Mathlib/Dynamics/Newton.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2024 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, Oliver Nash
+-/
+import Mathlib.Data.Polynomial.AlgebraMap
+import Mathlib.Data.Polynomial.Identities
+import Mathlib.RingTheory.Nilpotent
+import Mathlib.RingTheory.Polynomial.Nilpotent
+import Mathlib.RingTheory.Polynomial.Tower
+
+/-!
+# Newton-Raphson method
+
+Given a single-variable polynomial `P` with derivative `P'`, Newton's method concerns iteration of
+the rational map: `x ↦ x - P(x) / P'(x)`.
+
+Over a field it can serve as a root-finding algorithm. It is also useful tool in certain proofs
+such as Hensel's lemma and Jordan-Chevalley decomposition.
+
+## Main definitions / results:
+
+ * `Polynomial.newtonMap`: the map `x ↦ x - P(x) / P'(x)`, where `P'` is the derivative of the
+   polynomial `P`.
+ * `Polynomial.isFixedPt_newtonMap_of_isUnit_iff`: `x` is a fixed point for Newton iteration iff
+   it is a root of `P` (provided `P'(x)` is a unit).
+ * `Polynomial.exists_unique_nilpotent_sub_and_aeval_eq_zero`: if `x` is almost a root of `P` in the
+   sense that that `P(x)` is nilpotent (and `P'(x)` is a unit) then we may write `x` as a sum
+   `x = n + r` where `n` is nilpotent and `r` is a root of `P`. This can be used to prove the
+   Jordan-Chevalley decomposition of linear endomorphims.
+
+-/
+
+open Set Function
+
+noncomputable section
+
+namespace Polynomial
+
+variable {R S : Type*} [CommRing R] [CommRing S] [Algebra R S] (P : R[X]) {x : S}
+
+/-- Given a single-variable polynomial `P` with derivative `P'`, this is the map:
+`x ↦ x - P(x) / P'(x)`. When `P'(x)` is not a unit we use a junk-value pattern and send `x ↦ x`. -/
+def newtonMap (x : S) : S :=
+  x - (Ring.inverse <| aeval x (derivative P)) * (aeval x P)
+
+theorem newtonMap_apply :
```

Comment:
I think something like `newtonMap_apply_of_IsUnit` (or a similar name) is a reasonable addition, maybe even the case when `P'(x)` is not a unit.

Reply:
Good idea; done.